File: Benchmarking/stats/euler35/euler35.jl
#=
The number, 197, is called a circular prime because all rotations of the digits: 197, 971, and 719, are themselves prime.

There are thirteen such primes below 100: 2, 3, 5, 7, 11, 13, 17, 31, 37, 71, 73, 79, and 97.

How many circular primes are there below one million?
=#
#Answer: 55

#Terminal: julia euler35.jl 1000000
limit = parse(Int, ARGS[1])

function getPrimes(size)
	array = fill(true, size)
	array[1] = false
	c = 1
	for i = 2 : size
		if array[i] == true
			j = i + i
			while j <= size
				array[j] = false
				j += i
			end
			c += 1
		end
	end
	primeArray = Int[]
	for i = 2 : size
		if(array[i])
			push!(primeArray, i)
		end
	end
	return primeArray, array
end

function rotate(s) #Rotating n
  first = s[1:1]
  last = s[2:end]
  return string(last, first)
end

function calc(limit)
  #primes = array of primes
  #primesBool = array of booleans, if element is true, the index is a prime.
  #primesBool[11] = true, 11 is prime
  primes, primesBool = getPrimes(limit)

  count = 0 #Amount of circular primes found
  for p in primes
    s = string(p)
    l = length(s)
    circPrime = true
    for i = 1 : l-1
      s = rotate(s)
      n = parse(Int, s)
      if !primesBool[n]
        circPrime = false
        break
      end
    end
    if circPrime
      count += 1
    end
  end
  return count
end
calc(limit)

File: Benchmarking/todo/euler35/euler35.java
import java.util.*;

public class euler35
{
	public static void main(String[] args)
	{
		int limit = Integer.parseInt(args[0]);
		calc(limit);
	}
	
	private static PrimeResult getPrimes(int size)
	{
		ArrayList<Boolean> array = new ArrayList<Boolean>();
		for(int i = 0; i <= size; i++)
		{
			array.add(true);
		}
		array.set(1, false);
		int c = 1;
		for(int i = 2; i <= size; i++)
		{
			if(array.get(i))
			{
				int j = i + i;
				while(j <= size)
				{
					array.set(j, false);
					j += i;
				}
				c += 1;
			}
		}
		List<Integer> primeArray = new ArrayList<>();
		for(int i = 2; i <= size; i++)
		{
			if(array.get(i))
			{
				primeArray.add(i);
			}
		}
		return new PrimeResult(array, primeArray);
	}
	
	private static String rotate(String s)
	{
		String first = s.substring(0, 1);
		String last = s.substring(1, s.length());
		return last + first;
	}
	
	private static int calc(int limit)
	{
		PrimeResult pr = getPrimes(limit);
		ArrayList<Boolean> primesBool = pr.getBool();
		List<Integer> primes = pr.getPrimes();
		int count = 0;
		for(int p : primes)
		{
			String s = String.valueOf(p);
			int l = s.length();
			boolean circPrime = true;
			for(int i = 1; i < l; i++)
			{
				s = rotate(s);
				int n = Integer.parseInt(s);
				if(!primesBool.get(n))
				{
					circPrime = false;
					break;
				}
			}
			if(circPrime)
			{
				count += 1;
			}
		}
		return count;
	}
	
	private static class PrimeResult
	{
		ArrayList<Boolean> bools;
		List<Integer> primes;
		
		public ArrayList<Boolean> getBool()
		{
			return bools;
		}
		public List<Integer> getPrimes()
		{
			return primes;
		}		
		public PrimeResult(ArrayList<Boolean> b, List<Integer> p)
		{
			bools = b;
			primes = p;
		}
	}
}
